snap/errors.go

```diff
@@ -41,10 +45,43 @@ func (e NotInstalledError) Error() string {
 	return fmt.Sprintf("revision %s of snap %q is not installed", e.Rev, e.Snap)
 }
 
+// NotSnapError is returned if a operation expects a snap file or snap dir
+// but no valid input is provided.
 type NotSnapError struct {
 	Path string
+
+	Err error
 }
 
 func (e NotSnapError) Error() string {
-	return fmt.Sprintf("%q is not a snap or snapdir", e.Path)
+	return fmt.Sprintf("cannot process snap or snapdir: %v", e.Err)
+}
+
+func buildNotSnapErrorContext(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	if stat.IsDir() {
+		if _, err := f.Readdir(1); err == io.EOF {
+			return fmt.Errorf("directory %q is empty", path)
+		}
+		return fmt.Errorf("directory %q is invalid", path)
+	}
+
+	var header [10]byte
```

Comment:
where is the 10 coming from?

Reply:
The number is arbitrary, it could also be 5. It's just enough to give the user a chance to see that it's e.g. a PNG header instead of a squashfs header, i.e. give some debug hints.

Reply:
do we know how long is the normal squashfs marker?  anyway it probably merits a comment because other people might have the same question later

Reply:
Thanks, I added a comment.